### PR TITLE
 Order Fulfillment: Allow the user to enter package tracking details using a custom shipment provider

### LINKED ISSUE #846: Order Fulfillment: Allow the user to enter package tracking details using a custom shipment provider
Users should be able to create and assign a custom shipment provider to an order.

Screenshot 2019-04-09 at 3 39 16 PM

### PR DESCRIPTION
**Pull request by ctarda**
Fix #846 

This PR implements adding shipment tracking using a custom tracking provider. 

Simulator Screen Shot - iPhone 8 - 2019-04-18 at 14 31 35
Simulator Screen Shot - iPhone 8 - 2019-04-18 at 14 31 52

## Changes
- Networking: Added a parameter to the relevant method in `ShipmentsRemote` to pass the `shipmentDate`
- Yosemite: Added a parameter to the relevant methods in `ShipmentAction` and `ShipmentStore` to pass the `shipmentDate` when creating a shipment tracking with a custom provider
- `ManualTrackingViewModel`: Add an implementation for this use case, providing its own collection of sections to be rendered by the view controller. Also add and remove some properties to the interface.
- `ManualTrackingViewController`: Add configuration for cells corresponding to the new fields necessary to render the new sections, and add logic to dispatch the action to create a shipment tracking with a custom provider
- Unit tests and updates to the existing tests.

## Testing
- Checkout the branch, run the tests.
- Navigate to an order > Fulfill Order > Add Tracking
- Tap Shipping Provider (first row)
- Filter the list to reach the "error" view with a "Custom Provider" button. Tap the button
- Attempt to create a tracking with any name / tracking number.
- Name and tracking number are mandatory. URL is optional. If a url is entered, the system should  prepend `https://` is it's missing.
- It should still be possible to create tracking using one of the pre-defined providers
- In the Order Details screen, only trackings with url should show a "Track Package" button. 

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/ShipmentProvidersViewController.swift [hunk 1]
```diff
@@ -267,6 +267,10 @@ private extension ShipmentProvidersViewController {
     func addCustomProvider() {
         // TO be implemented as part of #846:
         // https://github.com/woocommerce/woocommerce-ios/issues/846
```
Comment (bummytime) [inline WooCommerce/Classes/ViewRelated/Orders/ShipmentProvidersViewController.swift]: Remove this comment?

**Comment by kyleaparker** [comment]
>I am wondering if we should prefill the provider name based on the search text? Just thinking out loud...it doesn't have to be done in this PR (cc @kyleaparker )

This sounds like a nice enhancement!

I might have missed something but from the gif it looks like the only way to select custom provider is through searching. Are we still going to add it to the list (under the store's country options)?

**Comment by ctarda** [comment]
Thanks folks. I just pushed a comment that:
- relies on the cell's default selection style (marking the background in grey)
- fixes the clipping 
- removes the unecessary comment 😊 

Added also issue #890 to pre-populate the name of the provider with the search term

**Comment by ctarda** [comment]
@kyleaparker My understanding (and I might very well be wrong) is that, if as [discussed in this p2 thread](https://wp.me/p99K0U-19V-p2), we do not store custom shipment providers, we would not be possible to add it to the list (under the country)

**Comment by loremattei** [comment]
Hey! I'm moving this PR to 1.8 since 1.7 has been cut. If you want it to make it to 1.7, feel free to target the release branch and to ping me to build a new 1.7 beta.

**Comment by kyleaparker** [comment]
>@kyleaparker My understanding (and I might very well be wrong) is that, if as discussed in this p2 thread, we do not store custom shipment providers, we would not be possible to add it to the list (under the country)

@ctarda that's correct but there should still be the option to select custom provider from that screen, we just don't store those providers:

Woo MVP Android-iOS i8 20190218 2019-04-25 12-07-11

Sorry for the misunderstanding, let me know if I should create an issue for this.

**Comment by ctarda** [comment]
@kyleaparker  Oh! I see! Sorry for the misunderstanding. I'll do that as part of #847 Thanks for pointing it out!

**Pull request merged**